### PR TITLE
Issue #1254 - Wallet not notified when transaction cannot be fulfilled

### PR DESCRIPTION
**Pull request by ferdeen**
Issue #1254.  Initial commit which includes an integration test around this issue.  

Moved ProofOfStakeMintCoinsSteps logic into a shared class, so it can be shared by other Integration tests that require coins to stake with.

### CODE REVIEW

File: src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs [hunk 1]
```diff
@@ -1,170 +1,73 @@
...
-            }
-
-            var coin = coins.First();
-            var txIn = transaction.AddInput(new TxIn(coin.Outpoint, this.powSenderAddress.ScriptPubKey));
-            transaction.AddOutput(new TxOut(new Money(9699999999995400), this.powSenderAddress.ScriptPubKey));
-            transaction.AddOutput(new TxOut(new Money(100000000000000), this.posReceiverAddress.ScriptPubKey));
-            transaction.Sign(this.powSenderPrivateKey, new[] { coin });
-
-            this.nodes[PowMiner].FullNode.NodeService<WalletController>().SendTransaction(new SendTransactionRequest(transaction.ToHex()));
+            this.proofOfStakeSteps.SendMillionCoinsFromPowWalletToPosWallet();
```
Comment (majikandy) [inline src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs]: perhaps SendOneMillionCoinsFromPowWalletToPosWalllet

File: src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs [hunk 2]
```diff
@@ -1,170 +1,73 @@
...
         private void pow_wallet_broadcasts_tx_of_million_coins_and_pos_wallet_receives()
         {
-            TestHelper.WaitLoop(() => this.nodes[PosStaker].CreateRPCClient().GetRawMempool().Length > 0);
-            TestHelper.WaitLoop(() => this.nodes[PosStaker].FullNode.WalletManager().GetSpendableTransactionsInWallet(PosWallet).Any());
-
-            this.sharedSteps.WaitForNodeToSync(this.nodes[PosStaker]);
-
-            var received = this.nodes[PosStaker].FullNode.WalletManager().GetSpendableTransactionsInWallet(PosWallet);
-            received.Sum(s => s.Transaction.Amount).Should().Be(Money.COIN * 1000000);
+            this.proofOfStakeSteps.PowWalletBroadcastsTransactionToPosWallet();
```
Comment (majikandy) [inline src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs]: which transaction?

File: src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs [hunk 3]
```diff
@@ -1,170 +1,73 @@
...
         }
 
         private void pos_node_mines_ten_blocks_more_ensuring_they_can_be_staked()
         {
-            this.nodes[PosStaker].GenerateStratisWithMiner(Convert.ToInt32(this.nodes[PosStaker].FullNode.Network.Consensus.Option<PosConsensusOptions>().CoinbaseMaturity));
+            this.proofOfStakeSteps.PosNodeMinesToEnsureStaking();
```
Comment (majikandy) [inline src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs]: this is more like  PosNodeMinesMinimumBlocksToAllowPreviousCoinsToBeStaked ......   but a better succinct version of that

File: src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs [hunk 4]
```diff
@@ -1,170 +1,73 @@
...
         }
 
         private void pos_node_starts_staking()
         {
-            var minter = this.nodes[PosStaker].FullNode.NodeService<IPosMinting>();
-            minter.Stake(new WalletSecret() { WalletName = PosWallet, WalletPassword = PosWalletPassword });
+            this.proofOfStakeSteps.PosNodeMinesToEnsureStaking();
```
Comment (majikandy) [inline src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs]: this is the same method call as the step above mining 10 blocks to allow staking, was that a typo ?

File: src/Stratis.Bitcoin.IntegrationTests/ProofOfStakeSteps.cs [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        private NodeGroupBuilder nodeGroupBuilder;
+        private SharedSteps sharedSteps;
+
+        private HdAddress powSenderAddress;
+        private Key powSenderPrivateKey;
+
+        private HdAddress posReceiverAddress;
+        private Key posReceiverPrivateKey;
+
+        public readonly string PowMiner = "ProofOfWorkNode";
```
Comment (majikandy) [inline src/Stratis.Bitcoin.IntegrationTests/ProofOfStakeSteps.cs]: should all these be constants? and why public ?  addressed from outside?

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/SendingRecentlyStakedCoinsBeforeMaturity.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+namespace Stratis.Bitcoin.IntegrationTests.Wallet
+{
+    public partial class SendingRecentlyStakedCoinsBeforeMaturity
+    {
+        [Fact]
+        public void Sending_Staked_Coins_Before_Maturity_Is_Not_Reflected_In_The_Wallets_Transaction_History()
+        {
+            Given(two_nodes_which_includes_a_proof_of_stake_node_with_a_million_coins);
+            When(a_wallet_sends_staked_coins_before_maturity);
+            Then(the_wallet_history_shows_the_transaction_as_sent);
```
Comment (majikandy) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/SendingRecentlyStakedCoinsBeforeMaturity.cs]: I can't see how this ties in with the title of the test.  Title says Not reflected in history, but the Then says it is.

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/SendingRecentlyStakedCoinsBeforeMaturity_Steps.cs [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+                .Sum(utxo => utxo.Transaction.Amount)
+                .Should().Be(0);
+        }
+
+        private WalletHistoryModel GetWalletHistory(CoreNode node, string walletName)
+        {
+            var walletHistory = node.FullNode.NodeService<WalletController>()
+                .GetHistory(new WalletHistoryRequest { WalletName = walletName }) as JsonResult;
+
+            return walletHistory.Value as WalletHistoryModel;
```
Comment (majikandy) [inline src/Stratis.Bitcoin.IntegrationTests/Wallet/SendingRecentlyStakedCoinsBeforeMaturity_Steps.cs]: happy with a null ref exception here or prefer a cast exception above?

File: src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs (hunk 1 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs]: Done - "SendOneMillionCoinsFromPowWalletToPosWallet".

File: src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs (hunk 2 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs]: Changed to PowWalletBroadcastsTransactionOfOneMillionCoinsAndPosWalletReceives.

File: src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs (hunk 3 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs]: Change to PosNodeMinesTenBlocksMoreEnsuringTheyCanBeStaked.

File: src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs (hunk 4 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/Miners/ProofOfStakeMintCoinsSteps.cs]: Good spot! Changed.

File: src/Stratis.Bitcoin.IntegrationTests/ProofOfStakeSteps.cs (hunk 5 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/ProofOfStakeSteps.cs]: Constants now and yes used outside the class.

Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/ProofOfStakeSteps.cs]: Addressed outside and need to be read only only fields as calling code returns "Member 'ProofOfStakeSteps.PosStaker' cannot be accessed with an instance reference; qualify it with a type name instead", when constants.

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/SendingRecentlyStakedCoinsBeforeMaturity.cs (hunk 6 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/Wallet/SendingRecentlyStakedCoinsBeforeMaturity.cs]: Changed to this "the_wallet_history_shows_the_transaction_as_sent_but_the_transaction_was_not_received".  Is this better ?

File: src/Stratis.Bitcoin.IntegrationTests/Wallet/SendingRecentlyStakedCoinsBeforeMaturity_Steps.cs (hunk 7 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.IntegrationTests/Wallet/SendingRecentlyStakedCoinsBeforeMaturity_Steps.cs]: Changed to return walletHistory?.Value as WalletHistoryModel;

**Comment by bokobza** [comment]
I'm going to review this as well.

**Comment by ferdeen** [comment]
The above changes fix the issue.  However, I need to refactor the code based on Dan's comments above.  I'm working on this now.

File: src/Stratis.Bitcoin/Broadcasting/TransactionBroadcastEntry.cs [hunk 8]
```diff
@@ -9,10 +9,13 @@ public class TransactionBroadcastEntry
 
         public State State { get; set; }
 
-        public TransactionBroadcastEntry(Transaction transaction, State state)
+        public string ErroMessage { get; set; }
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin/Broadcasting/TransactionBroadcastEntry.cs]: Typo error here

Comment (ferdeen) [reply src/Stratis.Bitcoin/Broadcasting/TransactionBroadcastEntry.cs]: Thanks

File: src/Stratis.Bitcoin.Features.Wallet/Broadcasting/FullNodeBroadcasterManager.cs [hunk 9]
```diff
@@ -29,8 +29,22 @@ public FullNodeBroadcasterManager(IConnectionManager connectionManager, IMempool
                 return;
 
             var state = new MempoolValidationState(false);
+
             if (!await this.mempoolValidator.AcceptToMemoryPool(state, transaction).ConfigureAwait(false))
-                this.AddOrUpdate(transaction, State.CantBroadcast);
+            {
+                var errorMessage = string.Empty;
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/Broadcasting/FullNodeBroadcasterManager.cs]: can be just `string errorMessage;`

File: src/Stratis.Bitcoin.Features.Wallet/Controllers/WalletController.cs [hunk 10]
```diff
@@ -715,10 +715,16 @@ public IActionResult SendTransaction([FromBody] SendTransactionRequest request)
                     });
                 }
 
-                this.walletManager.ProcessTransaction(transaction, null, null, false);
-
                 this.broadcasterManager.BroadcastTransactionAsync(transaction).GetAwaiter().GetResult();
 
+                var transactionBroadCastEntry = this.walletManager.GetTransactionBroadcastEntry();
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/Controllers/WalletController.cs]: I don't really like to see the walletManager responsible for returning a TrxBroasdcastEntry. What if there are many happening at the same time? Are we guaranteed that they will ever only be 1 to be concerned with?

At least I think we need to have a trx identifier and maybe this method can call  `GetTransaction(uint256 transactionHash)` from the `BroadcasterManagerBase`.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 11]
```diff
@@ -73,6 +73,9 @@ public class WalletManager : IWalletManager
         /// <summary>The broadcast manager.</summary>
         private readonly IBroadcasterManager broadcasterManager;
 
+        /// <summary>The transaction broadcast entry.</summary>
+        private TransactionBroadcastEntry transactionBroadcastEntry;
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: I think we wouldn't need this with my suggestion above.

File: src/Stratis.Bitcoin.Features.Wallet/Broadcasting/FullNodeBroadcasterManager.cs (hunk 9 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Wallet/Broadcasting/FullNodeBroadcasterManager.cs]: errorMessage is used a couple of lines down, so needs to be assigned at creation.

File: src/Stratis.Bitcoin.Features.Wallet/Controllers/WalletController.cs (hunk 10 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Wallet/Controllers/WalletController.cs]: Ok, didn't see this was available in BroadcasterManagerBase, thanks.

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs (hunk 11 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Cool.

**Comment by ferdeen** [comment]
Committed but need to run SendingStakedCoinsBeforeMaturity int test - which is currently breaking because of the int test file copy issue.  

Once I have merged the file copy issue fix - currently being reviewed now - from upstream master I'll check all is well against this PR.

**Pull request merged**